operator/pkg/watch/template_change.go

```diff
@@ -0,0 +1,120 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type TemplateChangeHandler struct {
+	client.Reader
+	client.StatusWriter
+	record.EventRecorder
+}
+
+func (h *TemplateChangeHandler) Watch(ctx context.Context) handler.MapFunc {
+	logger := log.FromContext(ctx).WithName("template-change-detection")
+	return func(o client.Object) []reconcile.Request {
+		requests := make([]reconcile.Request, 0)
+		template := &v1alpha1.ModuleTemplate{}
+
+		if err := h.Get(ctx, client.ObjectKeyFromObject(o), template); err != nil {
+			return requests
+		}
+
+		l := template.GetLabels()
+		managedBy, managedByPresent := l[labels.ManagedBy]
+		controller, controllerLabelPresent := l[labels.ControllerName]
+		channel, channelPresent := l[labels.Channel]
+		if !controllerLabelPresent || controller == "" ||
+			!channelPresent || channel == "" ||
+			!managedByPresent || managedBy != "kyma-operator" {
+			return requests
+		}
+
+		kymas := &v1alpha1.KymaList{}
+		err := h.List(ctx, kymas)
+		if err != nil {
+			return requests
+		}
+
+		templateNamespacedName := types.NamespacedName{
+			Namespace: template.GetNamespace(),
+			Name:      template.GetName(),
+		}
+		for _, kyma := range kymas.Items {
+			if kyma.Spec.Channel != v1alpha1.Channel(channel) {
+				continue
+			}
+
+			referenced := false
+			for _, component := range kyma.Spec.Components {
+				if component.Name == controller {
+					referenced = true
+					break
+				}
+			}
+			if !referenced {
+				continue
+			}
+
+			namespacedNameForKyma := types.NamespacedName{
+				Namespace: kyma.GetNamespace(),
+				Name:      kyma.GetName(),
+			}
+			logger.WithValues(
+				"controller", controller,
+				"channel", channel,
+				"template", templateNamespacedName.String(),
+			).Info(namespacedNameForKyma.String())
+
+			requests = append(requests, reconcile.Request{NamespacedName: namespacedNameForKyma})
+		}
+		return requests
+	}
+}
+
+func (h *TemplateChangeHandler) TemplateChange(
+	ctx context.Context,
+) func(event.UpdateEvent, workqueue.RateLimitingInterface) {
+	watch := h.Watch(ctx)
+	return func(event event.UpdateEvent, queue workqueue.RateLimitingInterface) {
+		requests := watch(event.ObjectNew)
+		for _, request := range requests {
+			queue.AddRateLimited(request)
+		}
+	}
+}
+
+func (h *TemplateChangeHandler) TemplateDelete(
+	ctx context.Context,
+) func(event.DeleteEvent, workqueue.RateLimitingInterface) {
+	watch := h.Watch(ctx)
+	return func(e event.DeleteEvent, queue workqueue.RateLimitingInterface) {
+		requests := watch(e.Object)
+		for _, request := range requests {
+			queue.AddRateLimited(request)
+		}
+	}
+}
+
+func (h *TemplateChangeHandler) TemplateCreate(
+	ctx context.Context,
+) func(event.CreateEvent, workqueue.RateLimitingInterface) {
+	watch := h.Watch(ctx)
+	return func(e event.CreateEvent, queue workqueue.RateLimitingInterface) {
+		requests := watch(e.Object)
+		for _, request := range requests {
+			queue.AddRateLimited(request)
+		}
+	}
+}
```

Comment:
are these functions used anywhere, maybe I missed it?

Reply:
These functions can be used as replacement of the mapfunc approach and I left them there to switch between the approaches for testing, will be removed